mipengine/node/monetdb_interface/monet_db_connection.py

```diff
@@ -1,117 +1,183 @@
+    """
+    MonetDBPool is a Singleton class because we want it to be initialized at runtime.
+
+    We use sudo-multithreading(eventlet greenlets),
+    we provide a connection pool to support concurrent query execution.
     """
 
     def __init__(self):
-        self._connection = None
-        self.refresh_connection()
         self._logger = logging.get_logger()
+        self._connection_pool = [self._create_connection() for _ in range(16)]
 
-    def refresh_connection(self):
-        self._connection = pymonetdb.connect(
+    def _create_connection(self):
+        return pymonetdb.connect(
             hostname=node_config.monetdb.ip,
             port=node_config.monetdb.port,
             username=node_config.monetdb.username,
             password=node_config.monetdb.password,
             database=node_config.monetdb.database,
         )
 
+    def _get_connection(self):
+        return self._connection_pool.pop()
+
+    def _release_connection(self, conn):
+        self._connection_pool.append(conn)
+
+    def _execute_and_commit(self, conn, db_execution_dto):
+        with self.cursor(conn) as cur:
+            cur.executemany(
+                db_execution_dto.query, db_execution_dto.parameters
+            ) if db_execution_dto.many else cur.execute(
+                db_execution_dto.query, db_execution_dto.parameters
+            )
+        conn.commit()
+
     @contextmanager
-    def cursor(self):
+    def cursor(self, _connection):
+        """
+        We use sudo-multithreading (eventlet greenlets) by committing before a select query we refresh the state
+        of the connection so that it sees changes from other connections.
+        https://stackoverflow.com/questions/9305669/mysql-python-connection-does-not-see-changes-to-database-made
+        """
+        _connection.commit()
 
-        broken_pipe_error = None
-        for _ in range(BROKEN_PIPE_MAX_ATTEMPTS):
-            try:
-                # We use a single instance of a connection and by committing before a select query we refresh the state
-                # of the connection so that it sees changes from other processes/connections.
-                # https://stackoverflow.com/questions/9305669/mysql-python-connection-does-not-see-changes-to-database-made
-                # -on-another-connect.
-                self._connection.commit()
-
-                cur = self._connection.cursor()
-                yield cur
-                cur.close()
-                break
-            except BrokenPipeError as exc:
-                broken_pipe_error = exc
-                self.refresh_connection()
-                continue
-            except Exception as exc:
-                raise exc
-        else:
-            raise broken_pipe_error
+        cur = _connection.cursor()
+        yield cur
+        cur.close()
 
-    def execute_and_fetchall(self, query: str, parameters=None, many=False) -> List:
+    @contextmanager
+    def lock(self, query_lock, timeout):
```

Comment:
`lock`, `execute` and `execute_and_fetchall` should be module methods, in my opinion.